SYCL/ESIMD/api/simd_memory_access.cpp

```diff
@@ -0,0 +1,181 @@
+    const uint32_t ii = static_cast<uint32_t>(i.get(0));
+    simd<T, N> v;
+    const auto offset = ii * sizeof(v);
+    v.copy_from(acc, offset);
+    v += simd<T, N>(ii * N, 1);
+    v.copy_to(acc, offset);
+  }
+};
+
+// Pointer-based kernel.
+template <typename T, int N> struct Kernel<T, N, false> {
+  T *ptr;
+  Kernel(T *ptr) : ptr(ptr) {}
+
+  void operator()(id<1> i) const SYCL_ESIMD_KERNEL {
+    using namespace sycl::INTEL::gpu;
+    const uint32_t ii = static_cast<uint32_t>(i.get(0));
+    simd<T, N> v;
+    const auto offset = ii * (sizeof(v) / sizeof(T));
+    v.copy_from(ptr + offset);
+    v += simd<T, N>(ii * N, 1);
+    v.copy_to(ptr + offset);
+  }
+};
+
+template <typename T> struct char_to_int {
+  using type = typename std::conditional<
+      sizeof(T) == 1,
+      typename std::conditional<std::is_signed<T>::value, int, unsigned>::type,
+      T>::type;
+};
+
+template <bool IsAcc, typename T> void free_mem(T *ptr, sycl::queue q) {
+  if constexpr (IsAcc) {
+    delete[] ptr;
+  } else {
+    sycl::free(ptr, q);
+  }
+}
+
+// The main test routine.
+template <typename T, int N, bool IsAcc> bool test(queue q, size_t size) {
+  std::cout << "Testing T=" << typeid(T).name() << ", N=" << N
+            << " using accessor=" << IsAcc << "...\n";
+  T *A;
+  if constexpr (IsAcc) {
+    A = new T[size];
+  } else {
+    A = reinterpret_cast<T *>(sycl::malloc_shared(size, q));
+  }
+
+  for (unsigned i = 0; i < size; ++i) {
+    A[i] = 0;
+  }
+
+  try {
+    if constexpr (IsAcc) {
+      buffer<T, 1> buf(A, range<1>(size));
+      range<1> glob_range{size / N};
+
+      auto e = q.submit([&](handler &cgh) {
+        auto acc = buf.template get_access<access::mode::read_write>(cgh);
+        Kernel<T, N, true> kernel(acc);
+        cgh.parallel_for(glob_range, kernel);
+      });
+    } else {
+      range<1> glob_range{size / N};
+
+      auto e = q.submit([&](handler &cgh) {
+        Kernel<T, N, false> kernel(A);
+        cgh.parallel_for(glob_range, kernel);
+      });
+    }
+    q.wait_and_throw();
+  } catch (cl::sycl::exception const &e) {
+    std::cout << "SYCL exception caught: " << e.what() << '\n';
+    free_mem<IsAcc>(A, q);
+    return e.get_cl_code();
+  }
+
+  int err_cnt = 0;
+
+  for (unsigned i = 0; i < size; ++i) {
+    T gold = (T)i;
+    T val = A[i];
+
+    if (val != gold) {
+      if (++err_cnt < 10) {
+        using T1 = typename char_to_int<T>::type;
+        std::cout << "failed at index " << i << ": " << (T1)val
+                  << " != " << (T1)gold << " (gold)\n";
+      }
+    }
+  }
+  if (err_cnt > 0) {
+    std::cout << "  pass rate: "
+              << ((float)(size - err_cnt) / (float)size) * 100.0f << "% ("
+              << (size - err_cnt) << "/" << size << ")\n";
+  }
+
+  free_mem<IsAcc>(A, q);
+
+  std::cout << (err_cnt > 0 ? "  FAILED\n" : "  Passed\n");
+  return err_cnt > 0 ? false : true;
+}
+
+int main(int argc, char **argv) {
+  size_t size = 32 * 7;
```

Comment:
Is it intentionally `32 * 7`, not `32 << 7` ?

Reply:
Yes